File: programmers/2021_kakao_internship/checking_distancing.py
from collections import deque


def check(row, col, place):
    direct_row = [-1, 0, 1, 0]
    direct_col = [0, -1, 0, 1]
    visited = {(row, col)}
    queue = deque()
    queue.append((row, col, 0))

    while queue:
        r, c, count = queue.popleft()

        if count == 2:
            continue

        for i in range(4):
            new_row, new_col = r + direct_row[i], c + direct_col[i]

            if 0 <= new_row < 5 and 0 <= new_col < 5 and (new_row, new_col) not in visited:
                if place[new_row][new_col] == "P":
                    return False
                elif place[new_row][new_col] == "X":
                    continue

                queue.append((new_row, new_col, count + 1))
                visited.add((new_row, new_col))

    return True


def solution(places):
    answer = []

    for place in places:
        candidates = []

        for i in range(len(place)):
            for j in range(len(place[i])):
                if place[i][j] == "P":
                    candidates.append((i, j))

        for row, col in candidates:
            if not check(row, col, place):
                answer.append(0)
                break
        else:
            answer.append(1)

    return answer


print(solution([["POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"], ["POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"],
                ["PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"], ["OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"],
                ["PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"]]))  # [1, 0, 1, 1, 1]

File: programmers/2021_kakao_internship/checking_distancing.js
const solution = places => {
    return places.map(place => {
        const rowLen = place.length;

        for (let i = 0; i < rowLen; i++) {
            // 열에 응시자가 없으면 다음 열로
            if (!place[i].includes('P')) {
                continue;
            }

            const columnLen = place[i].length;

            for (let j = 0; j < columnLen; j++) {
                if (place[i][j] === 'P' && ((
                    // 응시자 상하좌우에 다른 응시자가 있으면
                    (j > 0 && place[i][j - 1] === 'P') ||
                    (j < columnLen - 1 && place[i][j + 1] === 'P') ||
                    (i > 0 && place[i - 1][j] === 'P') ||
                    (i < rowLen - 1 && place[i + 1][j] === 'P')
                ) || (
                    // 응시자 대각선상에 다른 응시자가 있고, 대각선 좌우에 파티션이 하나라도 없다면
                    ((i < rowLen - 1 && j < columnLen - 1) && place[i + 1][j + 1] === 'P' && (place[i + 1][j] === 'O' || place[i][j + 1] === 'O')) ||
                    ((i < rowLen - 1 && j > 0) && place[i + 1][j - 1] === 'P' && (place[i + 1][j] === 'O' || place[i][j - 1] === 'O')) ||
                    ((i > 0 && j > 0) && place[i - 1][j - 1] === 'P' && (place[i - 1][j] === 'O' || place[j][j - 1] === 'O')) ||
                    ((i > 0 && j < columnLen - 1) && place[i - 1][j + 1] === 'P' && (place[i - 1][j] === 'O' || place[j][j + 1] === 'O'))
                ) || (
                    // 응시자와 한칸 떨어져있는 곳에 다른 응시자가 있고, 그 사이가 테이블이라면
                    (i < rowLen - 2 && place[i + 2][j] === 'P' && place[i + 1][j] === 'O') ||
                    (j > 1 && place[i][j - 2] === 'P' && place[i][j - 1] === 'O') ||
                    (i > 1 && place[i - 2][j] === 'P' && place[i - 1][j] === 'O') ||
                    (j < columnLen - 2 && place[i][j + 2] === 'P' && place[i][j + 1] === 'O')
                ))) {
                    return 0;
                }
            }
        }

        return 1;
    });
};
